helix-term/src/handlers/completion.rs

```diff
@@ -0,0 +1,465 @@
+    let config = editor.config.load();
+    if config.auto_completion {
+        let (view, doc): (&helix_view::View, &helix_view::Document) = current_ref!(editor);
+        let mut text = doc.text().slice(..);
+        let cursor = doc.selection(view.id).primary().cursor(text);
+        text = doc.text().slice(..cursor);
+
+        let is_trigger_char = doc
+            .language_servers_with_feature(LanguageServerFeature::Completion)
+            .any(|ls| {
+                matches!(&ls.capabilities().completion_provider, Some(lsp::CompletionOptions {
+                        trigger_characters: Some(triggers),
+                        ..
+                    }) if triggers.iter().any(|trigger| rope_ends_with(trigger, text)))
+            });
+        if is_trigger_char {
+            send_blocking(
+                tx,
+                CompletionEvent::TriggerChar {
+                    cursor,
+                    doc: doc.id(),
+                    view: view.id,
+                },
+            );
+            return;
+        }
+
+        let is_auto_trigger = !trigger_char_only
+            && doc
+                .text()
+                .chars_at(cursor)
+                .reversed()
+                .take(config.completion_trigger_len as usize)
+                .all(char_is_word);
+
+        if is_auto_trigger {
+            send_blocking(
+                tx,
+                CompletionEvent::AutoTrigger {
+                    cursor,
+                    doc: doc.id(),
+                    view: view.id,
+                },
+            );
+        }
+    }
+}
+
+fn update_completions(cx: &mut commands::Context, c: Option<char>) {
+    cx.callback.push(Box::new(move |compositor, cx| {
+        let editor_view = compositor.find::<ui::EditorView>().unwrap();
+        if let Some(completion) = &mut editor_view.completion {
+            completion.update_filter(c);
+            if completion.is_empty() {
+                editor_view.clear_completion(cx.editor);
+                // clearing completions might mean we want to immediately rerequest them (usually
+                // this occurs if typing a trigger char)
+                if c.is_some() {
+                    trigger_auto_completion(&cx.editor.handlers.completions, cx.editor, false);
+                }
+            }
+        }
+    }))
+}
+
+fn clear_completions(cx: &mut commands::Context) {
+    cx.callback.push(Box::new(|compositor, cx| {
+        let editor_view = compositor.find::<ui::EditorView>().unwrap();
+        editor_view.clear_completion(cx.editor);
+    }))
+}
+
+fn completion_post_command_hook(
+    tx: &Sender<CompletionEvent>,
+    PostCommand { command, cx }: &mut PostCommand<'_, '_>,
+) -> anyhow::Result<()> {
+    if cx.editor.mode == Mode::Insert {
+        if cx.editor.last_completion.is_some() {
+            match command {
+                MappableCommand::Static {
+                    name: "delete_word_forward" | "delete_char_forward" | "completion",
```

Comment:
I'm a bit worried about renaming functions potentially breaking this. Can we compare fn pointers instead?

Reply:
note that we are not matching random function names here but actual editor commands.

You can still rename the function without changing the command name. We conventionally name them the same but we don't have to. Renaming the commands is a breaking change. I would expect that we would eventually expose a similar event to plugins so renaming these commands would be both a breaking change to plugins and user configuration (which I don't really expect us to do or if we do it be very careful about).

I could also compare function pointers but sadly I cannot use them in the match so I would have to use a bunch of if/else chains which is a bit ugly. Most importantly this ties the handlers to helix-term which I try to avoid as much as possible (I want to move this kind of completion logic/handler to helix-view or another crate in the future since its not really terminal specific).

If you prefer I can still implement the `if fun == commands::delete_word_forward || fun == commands::delete_char_forward` solution but these were the reasons I didn't do so initially. What are your thoughts?

Reply:
Renaming the commands would be a breaking change and we aren't intending to do that, but I just worry that if that happens the two locations might end up out of sync. We already get various documentation issues because a keymap changed but wasn't updated in the docs but no lint caught that for example. Using pointers would ensure that it would be a compile time failure.

But I think you have a point and it isn't worth coupling the handler to helix-term